Fix dragLeave handling when children change (#370)

mwolson opened the pull request:
When the child content can vary according to whether the drag is active, such as putting an overlay in front of it or replacing some content, it's possible to get an initial dragEnter event when you drag onto that content, but you'll never get a corresponding dragLeave event, so the enter counter gets thrown off and the dragActive styles never go away.

* Track which elements have been the targets of `dragEnter` events
* On receiving a `dragLeave` event, remove from that list any DOM elements which are no longer part of the current container
* Add test case for this

okonet commented on src/index.js:
```diff
@@ -53,7 +53,8 @@ class Dropzone extends React.Component {
     // Count the dropzone and any children that are entered.
-    ++this.enterCounter;
+    this.dragTargets = this.dragTargets.filter(el => el !== e.target);
+    this.dragTargets.push(e.target);
```
I'm confused by this part TBH. You're filtering something and then you push to the array and mutate it. Can this be a single operation instead? `reduce` for example?

mwolson replied:
I'll change it to be:

```js
if (this.dragTargets.indexOf(e.target) === -1) {
  this.dragTargets.push(e.target);
}
```

okonet commented on src/index.js:
```diff
@@ -86,7 +87,9 @@ class Dropzone extends React.Component {
     // Only deactivate once the dropzone and all children was left.
-    if (--this.enterCounter > 0) {
+    this.dragTargets = this.dragTargets.filter(el => el !== e.target);
+    this.dragTargets = this.dragTargets.filter(el => this.node.contains(el));
```
Also here. This is very confusing to see 2 filters in a row with results overwriting each other. This can be rewritten like:

```
this.dragTargets =  this.dragTargets.filter(el => el !== e.target && this.node.contains(el));
```

but I'm wondering why do even need to assign it back to this instance variable if you're only concerned about the length of the array?

Wouldn't it easier to use `Array.prototype.some` or `Array.prototype.any` instead and not mutate it?

mwolson replied:
What you suggested seems fine, will implement:

```js
this.dragTargets =  this.dragTargets.filter(el => el !== e.target && this.node.contains(el));
```

okonet commented on src/index.js:
```diff
@@ -297,6 +300,7 @@ class Dropzone extends React.Component {
         onDragOver={this.onDragOver}
         onDragLeave={this.onDragLeave}
         onDrop={this.onDrop}
+        ref={node => (this.node = node)}
```
This will generate a new function each time. Instead, define a function outside of the scope and pass a reference to it.

mwolson replied:
OK.

mwolson commented:
Pushed the requested changes, ready for another round of review.

okonet commented on src/index.spec.js:
```diff
@@ -336,6 +336,7 @@ describe('Dropzone', () => {
       );
       dropzone.find(ChildComponent).simulate('dragEnter', { dataTransfer: { files } });
       dropzone.simulate('dragEnter', { dataTransfer: { files } });
+      dropzone.simulate('dragEnter', { dataTransfer: { files } }); // make sure we de-dupe events
```
Hmm, not sure I get this line. You're entering 2 times and leaving just 1 but asserts are passing? They shouldn't or am I missing something?

mwolson replied:
We want to make sure that even if the browser sends us duplicate `dragEnter` events for the same element, which it might, we'll still handle it correctly without the `dragTargets` list growing too large. It would also be possible to make the `dragEnter` just be a simple push and rely on the filter in `dragLeave` to remove any duplicates later, up to you.

okonet replied:
Makes sense! I think it's fine not to allocate too much memory and save some work later. Let's stick to it.

RaulTsc commented:
Let me know if you need any help with pushing this forward for a release.

okonet commented on src/index.js:
```diff
@@ -106,39 +106,47 @@ class Dropzone extends React.Component {
-      }
+    // Don't fire callbacks if a child Dropzone component has already handled this event
+    // eslint-disable-next-line no-underscore-dangle
+    if (!e.__dropzoneChildHandledThis) {
```
Why can't we use `event.stopPropagation` for this?

okonet commented on src/index.js:
```diff
@@ -106,39 +106,47 @@ class Dropzone extends React.Component {
+    // Don't fire callbacks if a child Dropzone component has already handled this event
+    // eslint-disable-next-line no-underscore-dangle
+    if (!e.__dropzoneChildHandledThis) {
+      e.persist();
```
Hmm, what this is doing?

mwolson commented:
Finished removing the handling of nested dropzone components and squashed the previous work. I added a test case to describe the current behavior, which is that both dropzones will fire onDrop() events.
